File: app/src/main/java/com/example/mike/mp3player/client/views/viewholders/MySongViewHolder.java
package com.example.mike.mp3player.client.views.viewholders;

import android.net.Uri;
import android.support.v4.media.MediaBrowserCompat;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.mike.mp3player.R;
import com.example.mike.mp3player.client.AlbumArtPainter;

import static com.example.mike.mp3player.commons.MediaItemUtils.extractArtist;
import static com.example.mike.mp3player.commons.MediaItemUtils.extractDuration;
import static com.example.mike.mp3player.commons.MediaItemUtils.extractTitle;
import static com.example.mike.mp3player.commons.MediaItemUtils.getAlbumArtUri;

public class MySongViewHolder extends MediaItemViewHolder {

    private final TextView title;
    private final TextView artist;
    private final TextView duration;
    private final ImageView albumArt;

    public MySongViewHolder(@NonNull View itemView, AlbumArtPainter albumArtPainter) {
        super(itemView, albumArtPainter);
        this.artist = itemView.findViewById(R.id.artist);
        this.title = itemView.findViewById(R.id.title);
        this.duration = itemView.findViewById(R.id.duration);
        this.albumArt = itemView.findViewById(R.id.song_item_album_art);
    }

    @Override
    public void bindMediaItem(MediaBrowserCompat.MediaItem item) {
        // - get element from your dataset at this position
        // - replace the contents of the views with that element
        String title = extractTitle(item);
        String artist = extractArtist(item);
        String duration = extractDuration(item);
        this.artist.setText(artist);
        this.title.setText(title);
        this.duration.setText(duration);
        Uri uri = getAlbumArtUri(item);
        albumArtPainter.paintOnView(albumArt, uri);
    }



}

File: client/src/main/java/com/github/goldy1992/mp3player/client/views/viewholders/MySongViewHolder.kt
package com.github.goldy1992.mp3player.client.views.viewholders

import android.net.Uri
import android.support.v4.media.MediaBrowserCompat
import android.support.v4.media.MediaMetadataCompat
import com.github.goldy1992.mp3player.client.AlbumArtPainter
import com.github.goldy1992.mp3player.client.databinding.SongItemMenuBinding
import com.github.goldy1992.mp3player.client.utils.TimerUtils.formatTime
import com.github.goldy1992.mp3player.commons.Constants
import com.github.goldy1992.mp3player.commons.MediaItemUtils
import com.github.goldy1992.mp3player.commons.MetaDataKeys
import org.apache.commons.io.FilenameUtils

class MySongViewHolder(private val view: SongItemMenuBinding, albumArtPainter: AlbumArtPainter?)
    : MediaItemViewHolder(view.root, albumArtPainter) {


    override fun bindMediaItem(item: MediaBrowserCompat.MediaItem) { // - get element from your dataset at this position
// - replace the contents of the views with that element
        val title : String? = extractTitle(item)
        val artist : String? = extractArtist(item)
        val duration : String? = extractDuration(item)
        view.artist.text = artist
        view.title.text = title
        view.duration.text = duration
        val uri : Uri? = MediaItemUtils.getAlbumArtUri(item)
        if (null != uri) {
            albumArtPainter!!.paintOnView(view.albumArt, uri)
        }
    }

    private fun extractTitle(song: MediaBrowserCompat.MediaItem): String {
        val charSequence: CharSequence? = MediaItemUtils.getTitle(song)
        return if (null == charSequence) {
            val fileName : String? = MediaItemUtils.getExtra(MetaDataKeys.META_DATA_KEY_FILE_NAME, song) as? String
            if (fileName == null)
                Constants.UNKNOWN else FilenameUtils.removeExtension(fileName)
            } else {
             charSequence.toString()
        }

    }

    private fun extractDuration(song: MediaBrowserCompat.MediaItem): String? {
        val extras = song.description.extras
        if (null != extras) {
            val duration = extras.getLong(MediaMetadataCompat.METADATA_KEY_DURATION)
            return formatTime(duration)
        }
        return null
    }

    private fun extractArtist(song: MediaBrowserCompat.MediaItem): String? {
        var artist: String?
        try {
            artist = MediaItemUtils.getArtist(song)
            if (null == artist) {
                artist = Constants.UNKNOWN
            }
        } catch (ex: NullPointerException) {
            artist = Constants.UNKNOWN
        }
        return artist
    }


}